Cubical/ZCohomology/RingStructure/GradedCommutativity.agda

```diff
@@ -0,0 +1,1082 @@
+-ₖ-helper {k = suc (suc k)} n m p q north = north
+-ₖ-helper {k = suc (suc k)} n m p q south = north
+-ₖ-helper {k = suc (suc k)} n m (inl x) q (merid a i) =
+  (merid a ∙ sym (merid (ptSn (suc k)))) i
+-ₖ-helper {k = suc (suc k)} n m (inr x) (inl x₁) (merid a i) =
+  (merid a ∙ sym (merid (ptSn (suc k)))) i
+-ₖ-helper {k = suc (suc k)} n m (inr x) (inr x₁) (merid a i) =
+  (merid a ∙ sym (merid (ptSn (suc k)))) (~ i)
+
+-ₖ-gen : {k : ℕ} (n m : ℕ)
+         (p : isEven' n ⊎ isOdd' n)
+         (q : isEven' m ⊎ isOdd' m)
+       → coHomK k → coHomK k
+-ₖ-gen {k = zero} = -ₖ-helper {k = zero}
+-ₖ-gen {k = suc k} n m p q = trMap (-ₖ-helper {k = suc k} n m p q)
+
+-- -ₖⁿ̇*ᵐ
+-ₖ^_·_ : {k : ℕ} (n m : ℕ) → coHomK k → coHomK k
+-ₖ^_·_ {k = zero} n m = -ₖ-helper {k = zero} n m (evenOrOdd n) (evenOrOdd m)
+-ₖ^_·_ {k = suc k} n m = trMap (-ₖ-helper n m (evenOrOdd n) (evenOrOdd m))
+
+-- cohomology version
+-ₕ^_·_ : {k : ℕ} {A : Type ℓ} (n m : ℕ) → coHom k A → coHom k A
+-ₕ^_·_ n m = sMap λ f x → (-ₖ^ n · m) (f x)
+
+-- -ₖⁿ̇*ᵐ = -ₖ for n m odd
+-ₖ-gen-inr≡-ₖ : {k : ℕ} (n m : ℕ) (p : _) (q : _) (x : coHomK k)
+  → -ₖ-gen n m (inr p) (inr q) x ≡ (-ₖ x)
+-ₖ-gen-inr≡-ₖ {k = zero} n m p q _ = refl
+-ₖ-gen-inr≡-ₖ {k = suc zero} n m p q =
+  trElim ((λ _ → isOfHLevelTruncPath))
+         λ { base → refl
+          ; (loop i) → refl}
+-ₖ-gen-inr≡-ₖ {k = suc (suc k)} n m p q =
+  trElim ((λ _ → isOfHLevelTruncPath))
+         λ { north → refl
+           ; south → refl
+           ; (merid a i) k → ∣ symDistr (merid (ptSn _)) (sym (merid a)) (~ k) (~ i) ∣ₕ}
+
+-- -ₖⁿ̇*ᵐ x = x for n even
+-ₖ-gen-inl-left : {k : ℕ} (n m : ℕ) (p : _) (q : _) (x : coHomK k)
+  → -ₖ-gen n m (inl p) q x ≡ x
+-ₖ-gen-inl-left {k = zero} n m p q x = refl
+-ₖ-gen-inl-left {k = suc zero} n m p q =
+  trElim (λ _ → isOfHLevelTruncPath)
+         λ { base → refl ; (loop i) → refl}
+-ₖ-gen-inl-left {k = suc (suc k)} n m p q =
+  trElim (λ _ → isOfHLevelPath (4 + k) (isOfHLevelTrunc (4 + k)) _ _)
+         λ { north → refl
+           ; south → cong ∣_∣ₕ (merid (ptSn _))
+           ; (merid a i) k → ∣ compPath-filler (merid a) (sym (merid (ptSn _))) (~ k) i ∣ₕ}
+
+-- -ₖⁿ̇*ᵐ x = x for m even
+-ₖ-gen-inl-right : {k : ℕ} (n m : ℕ) (p : _) (q : _) (x : coHomK k)
+  → -ₖ-gen n m p (inl q) x ≡ x
+-ₖ-gen-inl-right {k = zero} n m (inl x₁) q x = refl
+-ₖ-gen-inl-right {k = zero} n m (inr x₁) q x = refl
+-ₖ-gen-inl-right {k = suc zero} n m (inl x₁) q =
+  trElim (λ _ → isOfHLevelTruncPath)
+         λ { base → refl ; (loop i) → refl}
+-ₖ-gen-inl-right {k = suc zero} n m (inr x₁) q =
+  trElim (λ _ → isOfHLevelTruncPath)
+         λ { base → refl ; (loop i) → refl}
+-ₖ-gen-inl-right {k = suc (suc k)} n m (inl x₁) q =
+  trElim (λ _ → isOfHLevelTruncPath)
+         λ { north → refl
+           ; south → cong ∣_∣ₕ (merid (ptSn _))
+           ; (merid a i) k → ∣ compPath-filler (merid a) (sym (merid (ptSn _))) (~ k) i  ∣ₕ}
+-ₖ-gen-inl-right {k = suc (suc k)} n m (inr x₁) q =
```

Comment:
Is this this exactly the same as the `-ₖ-gen-inl-right {k = suc (suc k)} n m (inl x₁) q` case?

Reply:
Same for some other cases here... Why not use `_`?

Reply:
> Is this this exactly the same as the `-ₖ-gen-inl-right {k = suc (suc k)} n m (inl x₁) q` case?

Yeah, but I don't really see how I can reuse that proof. The two functions are only definitionally equal modulo pattern matching.

Reply:
> Same for some other cases here... Why not use `_`?

Where?